### PR TITLE
Bubble Sort Scratch Implementation

### PR DESCRIPTION
**Pull request by dovisutu**
In the building, I met a strange problem:
It can't pass. It said 
```
Template render error: (C:\Users\w\Downloads\algorithm-archive\contents\bubble_sort\bubble_sort.md) [Line 1605, Column 35]
parseSignature: expected comma after expression
```
what? Line 1605?
Also, I tried to paste the code of tree traversal here(with ../tree_traversal), but still fail.

### CODE REVIEW

**Comment by jiegillet** [comment]
Bubble sort usually sorts from smallest to largest, you should invert your logic.

In the error you showed there is a space in the path `contents\bubble_s ort\bubble_sort.md` is that a copy/paste error or did you include a space in the folder name?

**Comment by dovisutu** [comment]
That's an error when copying, because I directly copied from the console. No space in the folder.

**Comment by dovisutu** [comment]
Ok, I'll invert the logic(ascending)

File: contents/bubble_sort/bubble_sort.md [hunk 1]
```diff
@@ -141,6 +145,8 @@ Trust me, there are plenty of more complicated algorithms that do precisely the
 [import, lang:"emojicode"](code/emojicode/bubble_sort.emojic)
 {% sample lang="bash" %}
 [import, lang:"bash"](code/bash/bubble_sort.bash)
+{% sample lang="scratch"}
```
Comment (leios) [inline contents/bubble_sort/bubble_sort.md]: This should be `{% sample lang="scratch" %}`

File: contents/bubble_sort/bubble_sort.md [hunk 2]
```diff
@@ -68,6 +68,10 @@ This means that we need to go through the vector $$\mathcal{O}(n^2)$$ times with
 [import:2-14, lang:"emojicode"](code/emojicode/bubble_sort.emojic)
 {% sample lang="bash" %}
 [import:2-21, lang:"bash"](code/bash/bubble_sort.bash)
+{% sample lang="scratch" %}
+<p>
+  <img  class="center" src="/code/scratch/bubble_sort.svg" width="250" />
```
Comment (leios) [inline contents/bubble_sort/bubble_sort.md]: This should be `` (No `/` before the `code/` directory.

Also, I am not sure how you generated the image, but it is all black and looks weird.

Comment (jiegillet) [reply contents/bubble_sort/bubble_sort.md]: Yeah it's broken. The previous version looked OK (but code needed editing)

File: contents/bubble_sort/bubble_sort.md (hunk 1 above)
Comment (dovisutu) [reply contents/bubble_sort/bubble_sort.md]: sorry, a typo mistake...

File: contents/bubble_sort/bubble_sort.md (hunk 2 above)
Comment (dovisutu) [reply contents/bubble_sort/bubble_sort.md]: Sorry. I fixed these things (and build works).

**Comment by leios** [comment]
The scratch implementation that has been submitted is a png image that seems to have been saved as an svg. I don't mind png images in the algorithm archive, so feel free to use a png instead of svg. Could you also increase the quality of the image used?

**Comment by dovisutu** [comment]
Probably I can't increase the quality because I'm using `screenshot` in Windows, And what @jiegillet used to generate .svg directly from codeblock is now using `scratch 3.0` format (which is quite different from previous `scratch 2.0` format), but I'll still put it here as a compare.

s3blocks_github

**Comment by dovisutu** [comment]
wait, I just find the way to generate in `scratch 2`...crap.

**Comment by dovisutu** [comment]
OK, although I put a .svg generated from the tool @jiegillet used/

**Comment by leios** [comment]
All the scratch right now is in scratch2, but it is possible to output scratch3 with this: https://s3blocks.github.io/generator/ (use the export svg button). I would guess scratch3 would be better for this. You can get to this site by clicking the "scratchblocks" link in the upper left of the link you posted.

Would you also be able to commit the scratchblocks code used to generate this so we can generate it again in the future if there is a change to scratch (scratch4).

@jiegillet Is it possible to find the scratchblocks code you used for the scratch2 implementations?

**Comment by jiegillet** [comment]
Do you mean the code in text form? I'm afraid I dropped the ball and lost it, I should know better than to get rid of source code :(

**Comment by leios** [comment]
It's equally my fault for not asking for it at the time. I suppose we should use scratch3 for this one?

**Comment by jiegillet** [comment]
I think I deleted it even before sending the PR, nothing you could have done. I now think the code in text form should also be committed along with the svg (but should not be included in the `.md` file).

Yes, we should use Scratch 3 from now on! Maybe I'll update my Scratch svgs in the future!

**Comment by dovisutu** [comment]
well, how to include the scratchblock? just .txt?

**Comment by leios** [comment]
Yeah, please add a text document next to the svg image

**Comment by dovisutu** [comment]
Crap. I tried to generate in scratch 3, the picture does show on site, but when I click "Export SVG", nothing happened. What "Export PNG" makes a file that I can't open. When I directly save the picture, it becomes all black.
But it works when I save as PNG. Phew.
If anyone want the better `.svg` file, sorry, but I just can't make it here.

**Comment by leios** [comment]
I put a PR to your branch with the svg. It worked fine on my end.

**Comment by dovisutu** [comment]
@leios  sorry, but I found that I accidently added one more `define BubbleSort` line, and you're using that. Crap.

**Comment by leios** [comment]
I am not really sure why the svg option is not working for you

**Comment by dovisutu** [comment]
Phew. By some strange way, I finally get the correct svg picture.

**Comment by c252** [comment]
This isn't totally necessary, but it could be nice to change the variable names to match the other implementations ``` it -> i      it2 -> j    list -> a```

**Comment by dovisutu** [comment]
> This isn't totally necessary, but it could be nice to change the variable names to match the other implementations  it -> i      it2 -> j    list -> a

Thanks, but as mentioned before, I can't generate .svg here (may be caused by Edge), so I can't make change.
Although you can generate it and add it to here via PR.

**Comment by dovisutu** [comment]
hmm... I think I know how to make correct svg here.
Previously it was black, because this is missing:
```
<style><![CDATA[ .sb-label {
      font-family: "Helvetica Neue", Helvetica, sans-serif;
      font-weight: normal;
      fill: #fff;
      font-size: 11px;
      word-spacing: 0px;
      opacity: 1;
    }

    .sb-obsolete { fill: #ED4242; }
    .sb-motion { fill: #4C97FF; }
    .sb-looks { fill: #9966FF; }
    .sb-sound { fill: #CF63CF; }
    .sb-pen { fill: #0fBD8C;  }
    .sb-events { fill: #FFBF00; }
    .sb-control { fill: #FFAB19; }
    .sb-sensing { fill: #5CB1D6; }
    .sb-operators { fill: #59C059; }
    .sb-variables { fill: #FF8C1A; }
    .sb-list { fill: #FF661A }
    .sb-custom { fill: #FF6680; }
    .sb-custom-arg { fill: #FF6680; }
    .sb-oldextension { fill: #4b4a60; }
    .sb-extension { fill: #0fbd8c; }
    .sb-grey { fill: #969696; }
    .sb-bevel {
      filter2: url(#bevelFilter);
      stroke: #000;
      stroke-opacity: 0.15;
      stroke-alignment: inner;
    }
    .sb-input-round-dropdown,
    .sb-input-boolean {
      filter: url(#inputDarkFilter);
    }
    .sb-input {
      filter2: url(#inputBevelFilter);
      stroke: #000;
      stroke-opacity: 0.15;
      stroke-alignment: inner;
    }
    .sb-input-number,
    .sb-input-string,
    .sb-input-number-dropdown {
      fill: #fff;
    }
    .sb-literal-number,
    .sb-literal-string,
    .sb-literal-number-dropdown,
    .sb-literal-dropdown {
      font-weight: normal;
      font-size: 11px;
      word-spacing: 0;
    }
    .sb-literal-number,
    .sb-literal-string,
    .sb-literal-number-dropdown {
      fill: #444;
    }

    .sb-darker {
      filter2: url(#inputDarkFilter);
      stroke: #000;
      stroke-opacity: 0.1;
      stroke-alignment: inner;
    }
    .sb-desaturate {
      filter: url(#desaturateFilter);
    }
    .sb-outline {
      stroke: #000;
      stroke-opacity: 0.1;
      stroke-width: 1;
      fill: #FF4D6A;
    }

    .sb-define-hat-cap {
      stroke: #632d99;
      stroke-width: 1;
      fill: #8e2ec2;
    }

    .sb-comment {
      fill: #E4DB8C;
      stroke: #000;
      stroke-opacity: 0.2;
      stroke-width: 1;
    }
    .sb-comment-line {
      fill: #000;
      opacity: 0.2;
    }
    .sb-comment-label {
      font-family: "Helvetica Neue", Helvetica, sans-serif;
      font-weight: normal;
      fill: #000;
      font-size: 11px;
      word-spacing: 0px;
      opacity: 1;
    }

    .sb-diff {
      fill: none;
      stroke: #000;
    }
    .sb-diff-ins {
      stroke-width: 2px;
    }
    .sb-diff-del {
      stroke-width: 3px;
    }
  ]]></style>
```
so if I add it in the end of <def\> label, the image looks fine.
I think I *finally* can do the change requested by @c252 .

**Comment by leios** [comment]
I am relatively happy with the code and am almost ready to merge it. The only problem I have is that the BubbleSort function cannot take the list `a` as an input variable. This is a limitation with scratch, right?

If so, there is nothing we can do and I'm willing to merge now.

**Comment by dovisutu** [comment]
Yeah, we can't take lists as a variable.

**Pull request merged**